esp32/boards: Add Silicognition wESP32.

xorbit opened the pull request:
For rev 7+ boards with 16MB of flash.

xorbit committed:
esp32/boards: Add Silicognition wESP32.

For rev 7+ boards with 16MB of flash.

mattytrentini commented:
This looks good @xorbit, though I know that you have a small `boot.py` that you [ship with the boards](https://wesp32.com/software/#micropython) that enables the LAN port; you could add that to the build? If you'd like to do that, the [M5STACK_ATOM](https://github.com/micropython/micropython/tree/master/ports/esp32/boards/M5STACK_ATOM) or [UM_TINYPICO](https://github.com/micropython/micropython/tree/master/ports/esp32/boards/UM_TINYPICO) are examples that show how to freeze MicroPython files.

xorbit commented:
Thank you for the review @mattytrentini.  I was thinking about this but something is unclear to me.  If I understand correctly, frozen modules become built-in, there's no .py file for them in the file system.  So is it still possible for a user to override, replace, or expand this?
Or does a user supplied `boot.py` get executed in addition to the frozen one?
And would the `lan` object created in `boot.py` be available to the user?

xorbit commented:
Currently, I just load my `boot.py` in a separate step with `ampy`.  All it really does is configure and start the LAN connection for the board.
It sounds to me that it may be best to just keep doing that.  It keeps things transparent to the user.  They can look at the file and see what's going on, add to it if they like.  No "magic" going on. :)

So I think this PR can stay as is.

One more question though.  I was looking at the existing partition table `partitions-16MiB.csv` and it defines a weird size of 0xD59F80 for the `vfs` partition.  This seems to be the round value of 14000000 in decimal.  Not sure which board uses this.  In my new `partitions-16MiB-ota.csv` table I went with the fairly round-in-hex 0xcf0000 size, 13565952 in decimal.  I was just wondering if I'm missing something, if there's a good reason to do what's done in `partitions-16MiB.csv` that I'm not aware of or if it's just wasting 664kiB for the sake of "roundness" in decimal?

dpgeorge commented:
> It sounds to me that it may be best to just keep doing that. It keeps things transparent to the user.

I agree!

> One more question though. I was looking at the existing partition table `partitions-16MiB.csv` and it defines a weird size of 0xD59F80 for the `vfs` partition.

It does indeed.  It looks like I wrote that, and I really do not remember what is going on there.  So just ignore that.

dpgeorge commented on ports/esp32/partitions-16MiB-ota.csv:
```diff
@@ -0,0 +1,10 @@
+otadata,  data, ota,     0xd000,   0x2000,
+phy_init, data, phy,     0xf000,   0x1000,
+ota_0,    app,  ota_0,   0x10000,  0x180000,
+ota_1,    app,  ota_1,   0x190000, 0x180000,
```
I'd suggest making these partitions a bit larger.  The firmware size is already close to 1.5MiByte and the IDF does grow over time (we have no control over that).

xorbit replied:
I bumped the OTA partitions to 2 MiB, that should hopefully be good for a while.

xorbit replied:
Second guessing here.  Do you think that's big enough for the next couple of years at least?  Looks like about 0.5 MiB was gained from 1.10 to what will become 1.17.  That's from January 2019 till now, only 2.5 years.

xorbit committed:
Changed partition table to allocate 2 MiB for MicroPython

App partitions increased from 1.5 MiB to 2 MiB since MicroPython
is already 1.5 MiB and keeps growing.

xorbit committed:
Changed partition table to allocate 2.4 MiB for MicroPython

This will hopefully be sufficient for a long time.
Still 11 MiB left for the file system.

xorbit commented:
OK, bumped partition size for MicroPython to 2.4 MiB.  I'm confident that this should be sufficient for quite a while.  I'm ready to have this merged.

xorbit commented:
If there are no further issues, I would really like to see this merged before the 1.17 release.  My CM has started production on new boards that require new IDF 4.x based firmware and I would like to be able to give them an official release binary versus a dev build I do myself.

dpgeorge commented:
I see that the directory name of the new board is `SIL-WESP32`.  Do you mind changing this to `SIL_WESP32` to match all the other boards (that use underscore)?

xorbit commented:
Well that's just weird.  I tried to do this and found the resulting image now uses the default partition table, not mine.  After some checking I found that the generated `sdkconfig.combined` and `sdkconfig.combined.in` contain these two sections:

```
# For cmake build
CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions.csv"
```
and
```
# OTA

CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions-16MiB-ota.csv"
```

Seems the first `CONFIG_PARTITION_TABLE_CUSTOM_FILENAME` wins.  Which file is actually used to build the project?  `sdkconfig`, `sdkconfig.combined` or `sdkconfig.combined.in`?

xorbit committed:
Merge branch 'master' of github.com:micropython/micropython into wesp32

xorbit commented:
I even put my `sdkconfig.board` at the top of the list now in `mpconfigboard.cmake`, like so:
```
set(SDKCONFIG_DEFAULTS
    boards/SIL_WESP32/sdkconfig.board
    boards/sdkconfig.base
    boards/sdkconfig.ble
    boards/sdkconfig.240mhz
)

if(NOT MICROPY_FROZEN_MANIFEST)
    set(MICROPY_FROZEN_MANIFEST ${MICROPY_PORT_DIR}/boards/manifest.py)
endif()
```
This causes my `CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions-16MiB-ota.csv"` line to be first in the resulting `sdkconfig.combined` and `sdkconfig.combined.in`, before the errant `CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions.csv"` line.  But it doesn't help, the resulting image still uses the wrong partition table...

dpgeorge commented:
> Seems the first `CONFIG_PARTITION_TABLE_CUSTOM_FILENAME` wins. Which file is actually used to build the project? `sdkconfig`, `sdkconfig.combined` or `sdkconfig.combined.in`?

`sdkconfig.combined.in` is a concatenation of the input sdkconfig's defined by the board.  `sdkconfig.combined` is then a copy of that which is subsequently used to generate `sdkconfig`.  The latter is what is used to configure the build.

From my tests, duplicate config values that appear later in the `sdkconfig.combined` are the ones that are used (they override earlier ones).  Did you do a fresh build?  What IDF version, and what version of Python is used to build?

xorbit commented:
As for versions, this is a line I found in the build: 

```
-- Found PythonInterp: /home/xorbit/.espressif/python_env/idf4.4_py3.9_env/bin/python (found version "3.9.5")
```

I actually have IDF branch `release/v4.3` checked out at the moment because I wondered if being on `master` could cause a problem.

Anyway, today it works correctly.  I have no clue why it's different from yesterday, but today, the `sdkconfig` gets generated with 16MB flash and the correct partition table selected while yesterday it did not.  I did clean builds yesterday just as I did today as far as I know, my initial change of the board name MUST of course have triggered a clean build and it didn't work.  I know, something must have changed.  Maybe I didn't completely remove the build directory before I switched from IDF `master` to `release/v4.3`?  I may only have done a `make clean`.  Does that make a difference?

Anyway, I'll push this change since it seems to be working now.  It's a bit scary though, now knowing what caused it.  Probably should keep your eyes open when switching to newer IDF versions in the future to make sure.  Which IDF version is used for the official MicroPython builds?

xorbit committed:
esp32/boards: Changed board name as requested.

Changed board name from SIL-WESP32 to SIL_WESP32
to be consistent with other board names that use
underscores.

xorbit committed:
esp32/boards: Remove old board name.

Oops, failed to delete old SIL-WESP32 name in previous
commit.

dpgeorge commented:
> As for versions, this is a line I found in the build: ...

So that's IDF 4.4 with Python 3.9.  The reason the Python version may matter is because of ordering of elements in dicts, which could lead to duplicate config values having undefined preference.

> I may only have done a `make clean`. Does that make a difference?

`make clean` should be enough to delete and then get a fresh `sdkconfig` on the next build.

> Which IDF version is used for the official MicroPython builds?

4.2 for original ESP32 boards, and 4.3 for S2 and C3 boards.

dpgeorge commented:
Building locally with the latest master IDF 4.4, it works (the partition table is correct).

Squashed and merged in 028fc815cdb63b2ea50bc1f377177b19b6a3f9cc

xorbit commented:
Thanks Damien!

So will your build system now generate official release binaries for the wESP32?  If so, where will they be located?
How does a board get a specific download page like for instance the TinyPICO has?

dpgeorge commented:
> So will your build system now generate official release binaries for the wESP32?

For that there needs to be a line added to `tools/autobuild/build-esp32-latest.sh`.  Actually, it might be a good idea to just build all boards with some pattern matching logic in that file, similar to `tools/autobuild/build-stm32-latest.sh`.

> How does a board get a specific download page like for instance the TinyPICO has?

I have to make that manually... The plan is to eventually switch to something like CircuitPython has, where all boards have their own page with picture, description, URLs to outside websites, etc.  This would require some kind of metadata info in the board directory.

xorbit commented:
Oops, I guess an update to `tools/autobuild/build-esp32-latest.sh` should have been part of this PR.
Is that something I should do in a new PR or will you add it as part of the v1.17 release process?
